corehq/apps/api/resources/v1_0.py

```diff
@@ -49,3 +64,90 @@ def prepend_urls(self):
+    assigned_location_ids = fields.ListField(null=True)
+    profile = fields.CharField(null=True)
+    custom_user_data = fields.DictField(attribute='custom_user_data')
+    tableau_role = fields.CharField(attribute='tableau_role', null=True)
+    tableau_groups = fields.ListField(null=True)
+
+    class Meta(CustomResourceMeta):
+        resource_name = "invitation"
+        authentication = RequirePermissionAuthentication(HqPermissions.edit_web_users)
+        allowed_methods = ['post']
+        always_return_data = True
+
+    def dehydrate_role(self, bundle):
+        return bundle.obj.get_role_name()
+
+    def dehydrate_assigned_location_ids(self, bundle):
+        return [loc.location_id for loc in bundle.obj.assigned_locations.all() if loc is not None]
+
+    def dehydrate_primary_location_id(self, bundle):
+        if bundle.obj.primary_location:
+            return bundle.obj.primary_location.location_id
+
+    def dehydrate_tableau_groups(self, bundle):
+        return [group.name for group in get_tableau_groups_by_ids(bundle.obj.tableau_group_ids,
+                                                                 bundle.request.domain)]
+
+    def dehydrate_profile(self, bundle):
+        if bundle.obj.profile:
+            return bundle.obj.profile.name
+
+    def obj_create(self, bundle, **kwargs):
+        domain = kwargs['domain']
+        validator = WebUserResourceValidator(domain, bundle.request.couch_user)
+        spec = WebUserSpec(
+            email=bundle.data.pop('email'),
+            role=bundle.data.pop('role'),
+            primary_location_id=bundle.data.pop('primary_location_id', None),
+            assigned_location_ids=bundle.data.pop('assigned_location_ids', None),
+            profile=bundle.data.pop('profile', None),
+            custom_user_data=bundle.data.pop('custom_user_data', None),
+            tableau_role=bundle.data.pop('tableau_role', None),
+            tableau_groups=bundle.data.pop('tableau_groups', None),
+            unhandled_data=bundle.data,
+        )
+        errors = validator.is_valid(spec, True)
+        if errors:
+            raise ImmediateHttpResponse(JsonResponse({"errors": errors}, status=400))
+
+        profile = validator.profiles_by_name.get(spec.profile)
+        role_id = validator.roles_by_name.get(spec.role)
+        tableau_group_ids = get_tableau_group_ids_by_names(spec.tableau_groups or [], domain)
+
+        primary_loc = None
+        assigned_locs = []
+        if spec.assigned_location_ids:
+            primary_loc = SQLLocation.active_objects.get(
+                location_id=spec.primary_location_id)
+            assigned_locs = SQLLocation.active_objects.filter(
+                location_id__in=spec.assigned_location_ids, domain=domain)
+            real_ids = [loc.location_id for loc in assigned_locs]
+
+            if missing_ids := set(spec.assigned_location_ids) - set(real_ids):
+                raise ImmediateHttpResponse(JsonResponse(
```

Comment:
Could the location validation stuff go in the validator class?

Reply:
Yes it can, I decided to put it here to avoid having to fetch the location objects again. But maybe that extra fetch isn't a big deal, would definitely be cleaner to do the validation inside the validator class

Reply:
Ah, I definitely do support minimizing lookups, especially for an API, which can easily get spammed heavily.  The validator code already looks up locations, though I think it's pretty far deep - not sure how straightforward it'd be to get access to it - it's inside `validate_assigned_locations_allow_users`